change ddpLoggingData struct to map or dict

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54995 static graph api skeleton
* #54981 refactor autograd_hook
* #56730 add a test for grad view with torch amp
* **#56641 change ddpLoggingData struct to map or dict**

currently ddpLoggingData is flat struct, which requires internal DDP developers and external users to know about the struct field names. This is not flexible to delete or add new fields in the future. also it is hard to access ddpLoggingData.

With maps/dict, developers and users can easily access the fields without knowing the field names, also easier to add/remove a new/old field.

Since C++ does not support map values to be different types, right now ddpLoggingData containes two types of maps.

Differential Revision: [D27923723](https://our.internmc.facebook.com/intern/diff/D27923723/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27923723/)!

zhaojuanmao committed:
change ddpLoggingData struct to map or dict

currently ddpLoggingData is flat struct, which requires internal DDP developers and external users to know about the struct field names. This is not flexible to delete or add new fields in the future. also it is hard to access ddpLoggingData.

With maps/dict, developers and users can easily access the fields without knowing the field names, also easier to add/remove a new/old field.

Since C++ does not support map values to be different types, right now ddpLoggingData containes two types of maps.

Differential Revision: [D27923723](https://our.internmc.facebook.com/intern/diff/D27923723/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27923723/)!

[ghstack-poisoned]

SciPioneer commented on c10/util/Logging.cpp:
```diff
@@ -304,17 +303,16 @@ MessageLogger::MessageLogger(const char* file, int line, int severity)
       std::chrono::duration_cast<std::chrono::nanoseconds>(
           std::chrono::high_resolution_clock::now().time_since_epoch());
   */
```
nit: shall we delete this comment block?

SciPioneer commented on c10/util/Logging.cpp:
```diff
@@ -304,17 +303,16 @@ MessageLogger::MessageLogger(const char* file, int line, int severity)
-      << "] ";
+  stream_ << "["
+          << CAFFE2_SEVERITY_PREFIX[std::min(4, GLOG_FATAL - severity_)]
+          //<< (timeinfo->tm_mon + 1) * 100 + timeinfo->tm_mday
```
nit: shall we delete these comments?

SciPioneer commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1392,77 +1392,8 @@ Example::
-          "backward_compute_comm_overlap_time",
-          &c10::DDPLoggingData::backward_compute_comm_overlap_time)
-      .def_readwrite("is_multi_device_module", &c10::DDPLoggingData::is_multi_device_module);
+      .def_readwrite("strs_map", &c10::DDPLoggingData::strs_map)
```
You can also consider keeping a single map, where the value type is `c10::either<int, std::string>`.

Therefore, you can even just aliasing this class instead of creating a struct.
`using DDPLogingData = std::unordered_map<std::string, c10::either<int, std::string>>;`

SciPioneer replied:
You may also consider std::map instead. If the users don't know the field names and want to print them out, this can give a nicer and deterministic order. I don't think we need to care too much about the search performance here, since the map size is relatively small.

zhaojuanmao replied:
good suggestions, thanks!

zhaojuanmao replied:
> Optional: You can also consider keeping a single map, where the value type is `c10::either<int, std::string>`.
> 
> Therefore, you can even just aliasing this class instead of creating a struct.
> `using DDPLogingData = std::unordered_map<std::string, c10::either<int, std::string>>;`

just had second thoughts, c10::either will be hard to let us add more types in addition to int and string in the future. So I will keep current implementation for now

zhaojuanmao committed:
Update on "change ddpLoggingData struct to map or dict"


currently ddpLoggingData is flat struct, which requires internal DDP developers and external users to know about the struct field names. This is not flexible to delete or add new fields in the future. also it is hard to access ddpLoggingData.

With maps/dict, developers and users can easily access the fields without knowing the field names, also easier to add/remove a new/old field.

Since C++ does not support map values to be different types, right now ddpLoggingData containes two types of maps.

Differential Revision: [D27923723](https://our.internmc.facebook.com/intern/diff/D27923723/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27923723/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "change ddpLoggingData struct to map or dict"


currently ddpLoggingData is flat struct, which requires internal DDP developers and external users to know about the struct field names. This is not flexible to delete or add new fields in the future. also it is hard to access ddpLoggingData.

With maps/dict, developers and users can easily access the fields without knowing the field names, also easier to add/remove a new/old field.

Since C++ does not support map values to be different types, right now ddpLoggingData containes two types of maps.

Differential Revision: [D27923723](https://our.internmc.facebook.com/intern/diff/D27923723/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27923723/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "change ddpLoggingData struct to map or dict"


currently ddpLoggingData is flat struct, which requires internal DDP developers and external users to know about the struct field names. This is not flexible to delete or add new fields in the future. also it is hard to access ddpLoggingData.

With maps/dict, developers and users can easily access the fields without knowing the field names, also easier to add/remove a new/old field.

Since C++ does not support map values to be different types, right now ddpLoggingData containes two types of maps.

Differential Revision: [D27923723](https://our.internmc.facebook.com/intern/diff/D27923723/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27923723/)!

[ghstack-poisoned]

zhaojuanmao committed:
Update on "change ddpLoggingData struct to map or dict"


currently ddpLoggingData is flat struct, which requires internal DDP developers and external users to know about the struct field names. This is not flexible to delete or add new fields in the future. also it is hard to access ddpLoggingData.

With maps/dict, developers and users can easily access the fields without knowing the field names, also easier to add/remove a new/old field.

Since C++ does not support map values to be different types, right now ddpLoggingData containes two types of maps.

Differential Revision: [D27923723](https://our.internmc.facebook.com/intern/diff/D27923723/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D27923723/)!

[ghstack-poisoned]
